server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java

```diff
@@ -73,44 +75,55 @@
   private final ActiveRuleIndexer activeRuleIndexer;
   private final Languages languages;
   private final System2 system2;
+  private final DefaultOrganizationProvider defaultOrganizationProvider;
 
   public RegisterRules(RuleDefinitionsLoader defLoader, RuleActivator ruleActivator, DbClient dbClient, RuleIndexer ruleIndexer,
-    ActiveRuleIndexer activeRuleIndexer, Languages languages, System2 system2) {
+    ActiveRuleIndexer activeRuleIndexer, Languages languages, System2 system2, DefaultOrganizationProvider defaultOrganizationProvider) {
     this.defLoader = defLoader;
     this.ruleActivator = ruleActivator;
     this.dbClient = dbClient;
     this.ruleIndexer = ruleIndexer;
     this.activeRuleIndexer = activeRuleIndexer;
     this.languages = languages;
     this.system2 = system2;
+    this.defaultOrganizationProvider = defaultOrganizationProvider;
   }
 
   @Override
   public void start() {
     Profiler profiler = Profiler.create(LOG).startInfo("Register rules");
-    DbSession session = dbClient.openSession(false);
-    try {
+    try (DbSession session = dbClient.openSession(false)) {
       Map<RuleKey, RuleDefinitionDto> allRules = loadRules(session);
+      List<RuleKey> keysToIndex = new ArrayList<>();
 
       RulesDefinition.Context context = defLoader.load();
       for (RulesDefinition.ExtendedRepository repoDef : getRepositories(context)) {
         if (languages.get(repoDef.language()) != null) {
           for (RulesDefinition.Rule ruleDef : repoDef.rules()) {
-            registerRule(ruleDef, allRules, session);
+            boolean relevantForIndex = registerRule(ruleDef, allRules, session);
+            if (relevantForIndex) {
+              keysToIndex.add(RuleKey.of(ruleDef.repository().key(), ruleDef.key()));
+            }
           }
           session.commit();
         }
       }
-      List<RuleDefinitionDto> activeRules = processRemainingDbRules(allRules.values(), session);
-      List<ActiveRuleChange> changes = removeActiveRulesOnStillExistingRepositories(session, activeRules, context);
+      List<RuleDefinitionDto> removedRules = processRemainingDbRules(allRules.values(), session);
+      List<ActiveRuleChange> changes = removeActiveRulesOnStillExistingRepositories(session, removedRules, context);
       session.commit();
 
       persistRepositories(session, context.repositories());
-      ruleIndexer.index();
+      ruleIndexer.delete(removedRules.stream().map(r -> r.getKey()).collect(java.util.stream.Collectors.toList()));
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Replace this lambda with a method reference. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1612)